### PR TITLE
NativeAOT-LLVM: Merge up to feature/NativeAOT-final tag

### PR DESCRIPTION
**Pull request by yowl**
This PR moves the NativeAOT-LLVM branch forward to the https://github.com/dotnet/runtimelab/releases/tag/feature%2FNativeAOT-final tag.  From here we can merge direct from runtime (I hope).

WIP, to see what the non-wasm tests do..

cc @SingleAccretion

### CODE REVIEW

**Comment by jkotas** [comment]
`/eng/pipelines/common/global-build-job.yml (Line: 143, Col: 7): Unclosed function: 'and'. Located at position 1 within expression: and(eq(parameters.runtimeFlavor, 'coreclr'), contains(parameters.nameSuffix, 'Windows_wasm'). For more help, refer to https://go.microsoft.com/fwlink/?linkid=842996`

**Comment by yowl** [comment]
@jkotas  Thanks!

**Comment by jkotas** [comment]
> Error: `Could not find a pool with name NetCore1ESPool-Public. The pool does not exist or has not been authorized for use`.

I think it needs to be `NetCore-Public`

**Comment by SingleAccretion** [comment]
@yowl also, FYI, you can view the runtimelab pipeline here: https://dev.azure.com/dnceng-public/public/_build?definitionId=163.

File: src/coreclr/jit/llvmcodegen.cpp [hunk 1]
```diff
@@ -509,7 +509,7 @@ void Llvm::storeLocalVar(GenTreeLclVar* lclVar)
 // in case we haven't seen the phi args yet, create just the phi nodes and fill in the args at the end
 void Llvm::buildEmptyPhi(GenTreePhi* phi)
 {
-    llvm::PHINode* llvmPhiNode = _builder.CreatePHI(getLlvmTypeForVarType(phi->TypeGet()), phi->NumChildren());
+    llvm::PHINode* llvmPhiNode = _builder.CreatePHI(getLlvmTypeForVarType(phi->TypeGet()), 1 /* we have at least 1 (or 2?) value presumably */);
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvmcodegen.cpp]:
```suggestion
    llvm::PHINode* llvmPhiNode = _builder.CreatePHI(getLlvmTypeForVarType(phi->TypeGet()), 2);
```
`2` is a good estimate, 90% of PHIs are like that (and none will have `1`).
```
/// Constructors - NumReservedValues is a hint for the number of incoming
/// edges that this phi node will have (use 0 if you really have no idea).
```

File: src/coreclr/jit/target.h [hunk 2]
```diff
@@ -239,6 +239,9 @@ typedef unsigned __int64 regMaskSmall;
 typedef _regNumber_enum regNumber;
 typedef unsigned char   regNumberSmall;
 
+//TODO-LLVM: delete this when we are merged forward enough that it is no longer present
+#define SCALED_ADDR_MODES 1
+
```
Comment (SingleAccretion) [inline src/coreclr/jit/target.h]:
```suggestion
```
Deleted (in https://github.com/dotnet/runtime/pull/60808, which should be included in the commit stream here).

File: src/coreclr/jit/compiler.h [hunk 3]
```diff
@@ -8353,29 +8350,16 @@ class Compiler
 public:
     CodeGenInterface* codeGen;
 
-<<<<<<< HEAD
     //  The following holds information about instr offsets in terms of generated code.
```
Comment (SingleAccretion) [inline src/coreclr/jit/compiler.h]:
```suggestion
```
(This was for `IPmappingDsc`)

File: src/native/libs/CMakeLists.txt [hunk 4]
```diff
@@ -144,6 +148,7 @@ if (CLR_CMAKE_TARGET_UNIX OR CLR_CMAKE_TARGET_BROWSER)
         add_compile_options(-Wno-typedef-redefinition)
         add_compile_options(-Wno-c11-extensions)
         add_compile_options(-Wno-thread-safety-analysis)
+        add_compile_options(-Wno-strict-prototypes)
```
Comment (SingleAccretion) [inline src/native/libs/CMakeLists.txt]: Does it need a `TODO-LLVM` to fix the warnings?

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj [hunk 5]
```diff
@@ -39,13 +39,8 @@
...
     <NativeRuntimeAsset Include="$(CoreCLRCrossgen2Dir)$(LibPrefix)jitinterface_$(TargetArchitecture)$(LibSuffix)" TargetPath="tools/" />
     <NativeRuntimeAsset Include="$(CoreCLRCrossgen2Dir)$(LibPrefix)clrjit_win_x86_$(TargetArchitecture)$(LibSuffix)" TargetPath="tools/"  />
-<<<<<<< HEAD
-    <NativeRuntimeAsset Include="$(CoreCLRCrossgen2Dir)$(LibPrefix)clrjit_win_arm_$(TargetArchitecture)$(LibSuffix)" TargetPath="tools/"  />
-    <NativeRuntimeAsset Include="$(CoreCLRCrossgen2Dir)$(LibPrefix)clrjit_unix_arm_$(TargetArchitecture)$(LibSuffix)" TargetPath="tools/"  />
-    <NativeRuntimeAsset Condition="'$(TargetArchitecture)' == 'x64'" Include="$(CoreCLRCrossgen2Dir)$(LibPrefix)clrjit_browser_wasm32_$(TargetArchitecture)$(LibSuffix)"  TargetPath="tools/" />
-=======
     <NativeRuntimeAsset Include="$(CoreCLRCrossgen2Dir)$(LibPrefix)clrjit_universal_arm_$(TargetArchitecture)$(LibSuffix)" TargetPath="tools/"  />
->>>>>>> feature/NativeAOT-final
+    <NativeRuntimeAsset Condition="'$(TargetArchitecture)' == 'x64'" Include="$(CoreCLRCrossgen2Dir)$(LibPrefix)clrjit_browser_wasm32_$(TargetArchitecture)$(LibSuffix)"  TargetPath="tools/" />
```
Comment (SingleAccretion) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj]:
```suggestion
```
This is CG2 packaging. Does it need to include `wasmjit`?

(Presumably, we're not going to be R2R-ing to WASM any time soon)

**Review by SingleAccretion — COMMENTED** [review]
Looked at everything modulo `.yml`s.

File: src/coreclr/jit/target.h (hunk 2 above)
Comment (yowl) [reply src/coreclr/jit/target.h]: Deleted now

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj (hunk 5 above)
Comment (yowl) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj]: thanks

File: src/native/libs/CMakeLists.txt (hunk 4 above)
Comment (yowl) [reply src/native/libs/CMakeLists.txt]: Added a comment, and `-Wno-unused-but-set-variable` which was also failing in some pal files.

**Comment by yowl** [comment]
Aside from merge issues, the commit of note is https://github.com/dotnet/runtimelab/pull/2092/commits/81fd14d9303c36fe51e72e9fd5d90f454803c424 where we now allow the fallthrough with a StackValueType that is different.  This happens now in release config IL with e.g. ByRef and int.  Its in the IL->LLVM side, so I didn't think it was worth putting in a different check and just removed it completely.

**Comment by MichalStrehovsky** [comment]
> From here we can merge direct from runtime (I hope).

It might be worth to at least try and see if a subsequent merge from feature/NativeAOT proper wouldn't be helpful - those changes bridged over a couple diffs between how things were laid out here and how they ended up being laid out in dotnet/runtime main (off the top of my head, the runtime repo changes the parameters to be passed to `build.cmd` to build the nativeaot subset to `clr.nativeaotruntime+clr.nativeaotlibs` (and in later pull requests also adds a `clr.aot` macro).

If you go from runtime repo directly, you'll have to resolve those merge conflicts. If you go from feature/NativeAOT, I already resolved them.

**Comment by yowl** [comment]
> feature/NativeAOT proper

What do you mean by "proper" here?  I see that in runtimelab feature/NativeAOT has about another 45 days of commits.  If I go to the last commit in runtimelab feature/NativeAOT would that be a good next step?

**Review by jkotas — APPROVED** [review]
Nice!

**Comment by jkotas** [comment]
> If I go to the last commit in runtimelab feature/NativeAOT would that be a good next step?

Yes, I think so.

**Pull request merged**